src/reducer.ts

```diff
@@ -302,6 +310,18 @@ export default (oldState: State, action: Action): State => {
   return state
 }
 
+
+
+function deleteMessage (state: State, messageId: String) {
+  const target = state.messages.find(m => isDeletable(m) && m.messageId === messageId )
```

Comment:
I don't yet 100% understand the TS issue you mention without diving into the code in an editor myself, but: since message IDs should be unique, we can probably shrink down to one `isDeletable()` call by removing the one from the `find` function?

Reply:
Here's my understanding of what happens:

The `Message` class isn't required to have an `id` field; only Chat/Shout/Emote are. Therefore, you can't do `m.messageId === messageId` or even `m.messageId && m.messageId === messageId`. The `isDeletable` enforces the typing assurance of "the output of this function is in the class of Chat/Shout/Emote", which lets the `m.messageId === messageId` work.

However, the typing information doesn't get propagated out to `target`; target is of type `Message`. Therefore, we run into the same problem, except this time it's not on `messageId`, it's on `message` itself, since some message types (say, `MovedRoomMessage`) don't have a `message` field. So we need to again call the type guard function, so that `target.message = 'WORDS'` can be assured to be called on Chat/Shout/Emote.

You could solve this in a different way by explicitly casting the result of the find to `ChatMessage | ShoutMessage | EmoteMessage` but I thought it was better to just have it in the `isDeletable` function.